Add option to avoid cuda init to FakeTensorMode

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122668

Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

Differential Revision: [D55591785](https://our.internmc.facebook.com/intern/diff/D55591785)

zou3519 committed:
Add option to avoid cuda init to FakeTensorMode

Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

[ghstack-poisoned]

ezyang commented:
@eellison constant prop is always on tensors that are not too big, right? I suggest we just always do constant prop on CPU.

eellison commented:
@ezyang right it is always on tensor.numel() == 1

bdhirsh commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -869,23 +877,30 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
         maybe_prev_fake_mode = torch._C._unset_dispatch_mode(self._mode_key)
         if self is not maybe_prev_fake_mode:
-            self.enter_stack.append((True, maybe_prev_fake_mode))
+            self.enter_stack.append((True, maybe_prev_fake_mode, prev_lift_then_h2d))
```
Storing the config directly on FakeTensorMode does give us an annoying amount of flexibility when handling nested modes.

WDYT of just having the config be global state? Downside: you get weird behavior if you change the global state mid-way through the usage of your FakeTensorMode. Maybe that's a reason to stick with setting the config on FakeTensorMode directly, although the expected usage of the config is probably to globally enable it during all of export / model init.

zou3519 replied:
What's the problem with nested modes? We set and reset the state correctly on `__enter__/__exit__`.

bdhirsh replied:
I think I'm mostly just commenting on the fact that it feels unusual to do something like this:
```
with FakeTensorMode(prev_lift_then_h2d=True):
    foo()
    with FakeTensorMode(prev_lift_then_h2d=False):
        bar()
```

But we are giving users the full flexibility to let people do this, and we need to handle it via the extra state reseting on enter/exit.

I agree that this extra state resetting isn't actually that complicated. So if you want to keep the state on the mode then sgtm, I just figured I would call it out.

eellison replied:
See my other comment. Can we avoid adding this as a ctor and detect when it is necessary if cuda device is used and device is not avialable ?

bdhirsh commented on torch/csrc/utils/tensor_new.cpp:
```diff
@@ -453,19 +455,32 @@ Tensor internal_new_from_data(
-        device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/false);
+    if (!get_lift_then_h2d()) {
+      tensor = tensor.to(
+          device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/false);
```
Similar to my comment in `fake_tensor.py`, but: do we care about const-prop of dtype conversions still happening when this config is on? IIUC, we will still want to unconditionally do a `tensor.to(inferred_scalar_type)` to get that to work

zou3519 replied:
hmm, the config is supposed to change the ordering of `lift_fresh` w.r.t. to the tensor.to() call. Having something more fine-grained ("only do this for devices") adds more complexity to the code then I'd like ("lift then .to()" vs ".to() then lift" is easier to reason about).

If this is really a problem then we can make it more fine-grained in the future. The current solution doesn't prevent us from doing that.

I'll change the name of the config from `lift_then_h2d` to `lift_then_to` to better reflect the semantics

bdhirsh replied:
sgtm (agreed it is extra complexity, and I don't have any real signal on how important const-propping dtype conversions is)

zou3519 commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1282,6 +1297,15 @@ def maybe_to_constant(t):
+        # prop on constant tensors for ops that change devices.
+        avoiding_device_init = False
+        if self.avoid_device_init:
+            if func == torch.ops.aten._to_copy.default and "device" in kwargs:
```
yup

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -791,6 +791,7 @@ def __init__(
         allow_non_fake_inputs=False,
         shape_env=None,
         static_shapes=None,
+        avoid_device_init=False,
```
Do we need this as a flag ? Could we just set it unconditionally if we're trying to run cuda fake tensors with no cuda device available ?

eellison reviewed (commented):
Would be great to add support for constant prop, maybe as a separate pr on this stack.

bdhirsh reviewed (approved):
sgtm!

zou3519 committed:
Update on "Add option to avoid cuda init to FakeTensorMode"


Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55591785).

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -761,6 +761,7 @@ def __init__(
         allow_non_fake_inputs=False,
         shape_env=None,
         static_shapes=None,
+        _allow_unsafe_data_ptr_access=True,
```
Can we avoid adding this ctor option if this is not something we want around permanently ? Maybe take it from a config from functorch ?

zou3519 replied:
Yup

SherlockNoMad commented:
gentle ping for landing this.

zou3519 committed:
Update on "Add option to avoid cuda init to FakeTensorMode"


Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

Differential Revision: [D55591785](https://our.internmc.facebook.com/intern/diff/D55591785)

[ghstack-poisoned]

zou3519 committed:
Update on "Add option to avoid cuda init to FakeTensorMode"


Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

Differential Revision: [D55591785](https://our.internmc.facebook.com/intern/diff/D55591785)

[ghstack-poisoned]

zou3519 committed:
Update on "Add option to avoid cuda init to FakeTensorMode"


Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

Differential Revision: [D55591785](https://our.internmc.facebook.com/intern/diff/D55591785)

[ghstack-poisoned]

zou3519 committed:
Update on "Add option to avoid cuda init to FakeTensorMode"


Also partially fixes #122109

This PR:
- adds an option to avoid cuda init to FakeTensorMode (default: False).
  This is an option because changing this is BC-breaking and has perf
  implications.
- Concretely, the option (a) prevents constant propagation of ops that
  move the tensor to a new device and (b) changes the torch.tensor ctor
  strategy.
- We add a C++ flag (lift_then_h2d) to toggle the torch.tensor(1, device='cuda') ctor
  strategy. When false (default), it does the current PyTorch behavior
  of unconditionally constructing a concrete CUDA tensor then calling
  lift_fresh on it. When true, we instead construct a concrete CPU
  tensor, call lift_fresh, and then call Tensor.to (under any ambient
  modes).

Test Plan:
- existing tests

Differential Revision: [D55591785](https://our.internmc.facebook.com/intern/diff/D55591785)

[ghstack-poisoned]

zou3519 commented:
Summary of changes:
- we now auto-detect if we need to use the new option
- the new option is "construct the tensor on cpu, call lift_fresh, and do a final .to(device) later" instead of "construct the tensor on cuda and then call lift_fresh"

zou3519 commented:
@pytorchbot merge

SherlockNoMad commented:
land this via fbcode?

zou3519 commented:
This was landed in https://github.com/pytorch/pytorch/pull/124413
